File: 0797-all-paths-from-source-to-target/0797-all-paths-from-source-to-target.py
class Solution:
    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        return self.dfs(graph, 0, [0], [])

    def dfs(self, graph, i, curr_path, ans):
        if i == len(graph) - 1:
            ans.append(curr_path[:])
            return ans
        for node in graph[i]:
            curr_path.append(node)
            self.dfs(graph, node, curr_path, ans)
            curr_path.pop()
        return ans

File: 0797-all-paths-from-source-to-target/0797-all-paths-from-source-to-target.cpp
class Solution {
public:
    vector<vector<int>> allPathsSourceTarget(vector<vector<int>>& graph) {
        vector<vector<int>> ans;
        vector<int> path;
        dfs(0, graph, path, ans);
        return ans;
    }

private:
    void dfs(int node, const vector<vector<int>>& graph, vector<int>& path, vector<vector<int>>& ans) {
        path.push_back(node);

        if (node == graph.size() - 1) {
            ans.push_back(path);
        } else {
            for (auto neighbor : graph[node]) {
                dfs(neighbor, graph, path, ans);
            }
        }

        path.pop_back();
    }
};